File: projects/08-bloomFilter/implementation/bloomFilter.cpp
#include "murmur3.h"
#include <iostream>
#include <math.h>
#include <bitset>
#include <vector>
#include <string> 

using namespace std;

struct bloomFilter {

    // storage for the data structure
    vector<bool> bit;

    /* storage for seeds since cannot easily store hashing functions in cpp
     * doesn't increase size of implementation since you must store hashing functions to remain consistent
     * slight increases runtime since you must calculate the hashing function from the seed every time
     */
    vector<int> seeds;

    // number of hashing functions
    int k = 0;

    // size of BIT array
    int m;

    // expected number of elements to be added
    int n = 0;

    // defaults to making bloom filter with 1% false positive rate
    bloomFilter(int n) {
        this->m = this->getMByP(0.01, n);
        this->n = n;
        this->bit.assign(this->m, false);
        k = this->optimalK(this->m, n);
        setSeeds(k);
    }

    // creates bloomFilter at specific P
    bloomFilter(float P, int n) {
        this->m = this->getMByP(P, n);
        this->n = n;
        this->bit.assign(this->m, false);
        this->k = this->optimalK(this->m, n);
        setSeeds(this->k);
    }

    // creates a bloomFilter of size m
    bloomFilter(int m, int n) {
        this->m =  m;
        this->n = n;
        this->bit.assign(this->m, false);
        this->k = this->optimalK(this->m, n);
        setSeeds(this->k);
    }

    // sets k seed values to random numbers
    void setSeeds(int k) {
        this->k = k;
        srand(time(0));
        for(int i = 0; i < k; i++) {
            this->seeds.push_back(rand());
        }
    }

    // calculates the probability of false positive
    int getProb() {
        return (int)pow(1 - pow((1 - (1 / m)), k * n), k);
    }

    // sets m to guarantee a probability P on n inputs
    int getMByP(double P, int n) {
        return (int)-((n * log(P)) / pow(log(2), 2));
    }

    // sets m to any integer
    void setM(int m) {
        this->m = m;
    }

    // sets n to any integer
    void setN(int n) {
        this->n = n;
    }

    // returns optimal number of hashing functions given n and m
    int optimalK(int m, int n) {
        return (int)((this->m / this->n) * log(2));
    }

    // method to add an int
    void add(void *n) {
        for (int i = 0; i < k; ++i) {
            int s = seeds[i];
            uint32_t h = hash(n, s); // gets k hashes and sets relevant indices to true in bit
            bit[h % m] = true;
        }
    }

    // wrapper for int hashing
    uint32_t hash(void *n, int seed) {
        void *i = &n;
        uint32_t o[4];
        MurmurHash3_x86_32(n, 8, seed, o);
        return *(uint32_t*)o; 
    }

    // checks if bf contains a value
    bool contains(void *n) {
        for(int i = 0; i < k; i++) {
            int s = seeds[i];
            uint32_t h = hash(n, s);
            if (bit[h % m] == false) { // if any index is false, return early
                return false;
            }
        }
        return true; // if all hashed indices are true, return true
    }
};

int main() {
    long numToAdd;
    long numToTest;

    cin >> numToAdd;
    cin >> numToTest;

    bloomFilter bf = bloomFilter(numToAdd);
    
    // reads in and adds values to bloom filter
    for (long i = 0; i < numToAdd; i++) {
        long toAdd;
        cin >> toAdd;
        bf.add(&toAdd);
    }

    int in = 0;

    // checks if test object is in the bloom filter
    for (long i = 0; i < numToTest; i++) {
        long toTest;
        cin >> toTest;
        if (bf.contains(&toTest))
            in++;
    }

    cout << "in: " << in << '\n';
    cout << "out: " << numToTest - in << '\n';
    return 0;
}

File: projects/08-bloomFilter/implementation/bloomFilter.py
import math
import mmh3
import random
import threading
import struct

class BloomFilter:
    # defaults to making bloom filter with 1% false positive rate
    def __init__(self, n, P=0.01, m=None):
        """
        n: Expected number of elements.
        m: Size of bit array. If None, computed from false positive
           probability P using m = -(n * ln(P)) / (ln2)^2.
        P: Target false positive probability.
        bit: Bloom Filter array representation
        k: Number of hash functions
        """
        if m is None:
            self.m = self.get_m_by_p(P, n) # makes bloomFilter size depending on P and n
        else:
            self.m = m # creates a bloomFilter of size m
        self.n = n
        self.bit = [False] * self.m
        self.k = self.optimal_k(self.m, self.n)
        self.set_seeds(self.k)

    def set_m(self, m):
        self.m = m

    def set_n(self, n):
        self.n = n
        
    # sets m to guarantee a probability P on n inputs
    def get_m_by_p(self, P, n):
        return int(-(n * math.log(P)) / math.pow(math.log(2), 2))

    # returns optimal number of hashing functions given n and m
    def optimal_k(self, m, n):
        return int((m / n) * math.log(2))

    # sets k seed values to random numbers
    def set_seeds(self, k):
        self.k = k
        rng = random.Random()
        self.seeds = [rng.randint(0, 2147483647) for _ in range(k)]

    # insertion for a string
    def add_string(self, s):
        for i in range(self.k): # go through all the hash functions
            h = self.hash_string(s, self.seeds[i]) # get the index from the current hash function
            self.bit[h % self.m] = True # change the corresponding bit in the bit array to be True

    # wrapper for string hashing
    def hash_string(self, s, seed):
        byte_array = s.encode('utf-8') # Convert string to bytes
        return mmh3.hash(byte_array, seed) & 0xffffffff # call mmh3 hash function and then do a bit shift to ensure non-negative

    # insertion for a string
    def add_int(self, n):
        for i in range(self.k):
            h = self.hash_int(n, self.seeds[i])
            self.bit[h % self.m] = True

    # wrapper for int hashing
    def hash_int(self, n, seed):
        big_int_n = int(n)
        byte_array = self.to_byte_array(big_int_n) # convert to bytes
        return mmh3.hash(byte_array, seed) & 0xffffffff # call mmh3 hash function and then do a bit shift to ensure non-negative

    # convert an int to a byte array
    def to_byte_array(self, num): # https://stackoverflow.com/questions/23870859/tobytearray-in-python
        bytea = []
        n = num
        while n:
            bytea.append(n % 256)
            n //= 256
        n_bytes = len(bytea)
        if 2 ** (n_bytes * 8 - 1) <= num < 2 ** (n_bytes * 8):
            bytea.append(0)
        return bytearray(reversed(bytea))

    # query element membership in Bloom Filter
    def contains(self, s):
        for i in range(self.k): # go through k hash filters
            h = self.hash_string(s, self.seeds[i])  # get index by hashing element for current hash "seed"
            if not self.bit[h % self.m]:
                return False
        return True # element only in filter if every index mapped by the hashing functions has been set to True

def main():
    num_inputs, num_checks = map(int, input().split())
    bf = BloomFilter(num_inputs)

    # input
    for _ in range(num_inputs):
        s = input().strip()
        bf.add_string(s)

    in_ = 0

    # output
    for _ in range(num_checks):
        s = input().strip()
        if bf.contains(s):
            in_ += 1

    print(f"In: {in_}")
    print(f"Out: {num_checks - in_}")

if __name__ == "__main__":
    main()